api/src/main/java/com/google/sps/SendNotifications.java

```diff
@@ -1,117 +1,155 @@
 package com.google.sps;
 
-import javax.servlet.annotation.WebServlet;
-import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-import com.google.auth.oauth2.GoogleCredentials;
 import com.google.cloud.firestore.CollectionReference;
 import com.google.cloud.firestore.DocumentReference;
 import com.google.cloud.firestore.DocumentSnapshot;
 import com.google.cloud.firestore.Firestore;
-import com.google.cloud.firestore.QueryDocumentSnapshot;
-import com.google.cloud.firestore.QuerySnapshot;
-import com.google.firebase.FirebaseApp;
-import com.google.firebase.FirebaseOptions;
 import com.google.firebase.cloud.FirestoreClient;
 import com.google.api.services.gmail.model.Message;
-import com.google.api.core.ApiFuture;
 import com.google.api.services.gmail.Gmail;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.util.Properties;
+import java.util.concurrent.ExecutionException;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
-import java.util.Map;
 
 import javax.mail.MessagingException;
 import javax.mail.Session;
 import javax.mail.internet.*;
 
 import org.apache.commons.codec.binary.Base64;
-import org.springframework.boot.autoconfigure.SpringBootApplication;
-import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 public class SendNotifications {
+  static boolean firebaseInitialized = false;
```

Comment:
I don't feel it is a good pattern. We can initialize the firebase once in our main() function before our server starts to receive requests. And in request handling, we're sure firebase has been correctly setup.

The current initialization with multiple places calling Firebase.init() makes it hard to reason the initialization sequence. Try to answer this question, if Firebase.init() cannot succeed (for example, we give a bad format credential file), when will the server fail, during startup time or when handling requests? A server should try its best to make sure all necessary resources are ready before it starts receiving requests.